byterun/str.c

```diff
@@ -411,7 +411,7 @@ CAMLexport value caml_alloc_sprintf(const char * format, ...)
        Note that caml_alloc_string left room for a '\0' at position n,
        so the size passed to vsnprintf is n+1. */
     va_start(args, format);
-    vsnprintf(String_val(res), n + 1, saved_format, args);
+    vsnprintf((char *)String_val(res), n + 1, saved_format, args);
```

Comment:
Do you have in mind something like this?
```C
value caml_alloc_writeable_string(mlsize_t len, char **p);
```
If so, that seems less clear to me, as returning multiple values in C is awkward:
```C
char *p;
value v = caml_alloc_writeable_string(len, &p);
vsnprintf(p, …)
```
But addressing the allocate + initialize pattern is a good idea.  I've added a new function `caml_alloc_initialized_string` (fd7fb761) that addresses the majority of these cases, where the string is initialized from a length and a source address.  Here's a typical change, from `byterun/str.c`:
```diff
-    res = caml_alloc_string(n);
-    memcpy((char *)String_val(res), buf, n);
+    res = caml_alloc_initialized_string(n, buf);
```

With this new function the length, which was previously passed to both `caml_alloc_string` and `memcpy`, is no longer repeated; there's no cast or call to `String_val` in the caller, and the caller never needs to touch a writeable string.  What do you think?